### ISSUE TITLE
Return struct errors compatible with `encoding/json`

### ISSUE THREAD

**Issue by luxas** [json-iterator/go#572]
Hi, thanks for this nice project!

It would be really nice if this project would return errors (exactly or wrapping) e.g. [`json.UnmarshalTypeError`](https://pkg.go.dev/encoding/json#UnmarshalTypeError) or [`json.SyntaxError`](https://pkg.go.dev/encoding/json#SyntaxError), for even more compatibility with the stdlib.

Showing the context like this project already does now is still nice, hence I think it's acceptable to return an error wrapping the above mentioned structs and implementing the `errors.Is/As` interfaces for equality checking.

I would be happy to contribute this if accepted :+1:

**Comment by iambudi** [comment]
Agree. It's nice to get this implemented. I have to go back to the standard json decoder because of this.

**Comment by adirizka7** [comment]
bringing this up 🆙

**Comment by SyntaxNode** [comment]
We were evaluating this library and got excited for `100% compatible drop-in replacement of "encoding/json"`. Unfortunately, we found a lot of cases where that statement doesn't hold up, this being one of them.
